### DISCUSSION TITLE
npm install して脆弱性が検出された時、どうしたら良いか

### DISCUSSION THREAD

**Discussion by kagomen** [lgtm-factory/lgtm-factory#67] [Scrap]
```sh
npm i -D prettier
```
上記を実行したら、脆弱性が検出された

```sh
added 1 package, and audited 623 packages in 724ms

5 vulnerabilities (1 low, 2 moderate, 2 high)

To address all issues, run:
  npm audit fix

Run `npm audit` for details.
```

以下を実行

```sh
npm audit fix
```

しかし改善されてない模様

```sh
up to date, audited 623 packages in 1s

150 packages are looking for funding
  run `npm fund` for details

# npm audit report

debug  4.0.0 - 4.3.0
Regular Expression Denial of Service in debug - https://github.com/advisories/GHSA-gxpj-cx7g-858c
fix available via `npm audit fix`
node_modules/@vercel/fun/node_modules/debug
  @vercel/fun  *
  Depends on vulnerable versions of debug
  Depends on vulnerable versions of semver
  Depends on vulnerable versions of tar
  node_modules/@vercel/fun
    vercel  >=32.3.1
    Depends on vulnerable versions of @vercel/fun
    node_modules/vercel

semver  7.0.0 - 7.5.1
Severity: high
semver vulnerable to Regular Expression Denial of Service - https://github.com/advisories/GHSA-c2qf-rxjj-qqgw
fix available via `npm audit fix`
node_modules/@vercel/fun/node_modules/semver

tar  <6.2.1
Severity: moderate
Denial of service while parsing a tar file due to lack of folders count validation - https://github.com/advisories/GHSA-f5x3-32g6-xq36
fix available via `npm audit fix`
node_modules/tar

5 vulnerabilities (1 low, 2 moderate, 2 high)

To address all issues, run:
  npm audit fix
```

**Comment by kagomen** [comment]
はい、その理解で正確です。

1. `npm audit fix` を実行しましたが、すべての脆弱性が修正されませんでした。

2. 再度 `npm audit fix` を実行することは推奨されていますが、結果が変わる可能性は低いです。なぜなら：

   - 自動的に修正できない脆弱性が残っている可能性が高いです。
   - これらの脆弱性は、依存関係の制約や互換性の問題により、簡単には修正できない場合があります。

このような状況では、以下のアプローチを検討することをお勧めします：

1. まず、`npm audit fix` をもう一度実行してみてください。変化がない可能性が高いですが、確認する価値はあります。

2. より積極的な修正を試みる場合は、`npm audit fix --force` を使用できます。ただし、これは互換性の問題を引き起こす可能性があるので、注意が必要です。

3. 個々の脆弱性を手動で調査し、対処することを検討してください：
   - 報告された各脆弱性の重要度を評価します。
   - 影響を受けるパッケージの最新バージョンをチェックし、手動でアップデートすることを検討します。
   - プロジェクトの他の依存関係との互換性を確認しながら、慎重にアップデートを行います。

4. 一部の脆弱性が長期的に解決できない場合、そのリスクを受け入れるか、影響を受けるパッケージの使用を避ける方法を検討する必要があるかもしれません。

5. 定期的に `npm audit` を実行し、新しい更新や修正が利用可能になったら適用することをお勧めします。

重要なのは、脆弱性の重要度と、それらがプロジェクトに与える潜在的な影響を理解することです。すべての脆弱性を即座に修正する必要はありませんが、高リスクの脆弱性には優先的に対処すべきです。

**Reply by kagomen** [reply]
```
npm audit fix --force
```

を実行してみたが、何も変わらなかった

**Comment by kagomen** [comment]
空のディレクトリで、

```sh
npm init -y
npm i -D prettier
```
を実行してみたら以下のように、脆弱性の問題は出てこなかった

```
added 1 package, and audited 2 packages in 485ms

1 package is looking for funding
  run `npm fund` for details

found 0 vulnerabilities
```

Prettierに脆弱性あるんじゃなくて、依存関係を解消する？ライブラリに脆弱性がある、ということなのかー

**Reply by kagomen** [reply]
その通りです。あなたの観察は正確です。この状況から以下のことが分かります：

1. Prettier自体には脆弱性がない：
   新しく作成したプロジェクトにPrettierをインストールした際、脆弱性は報告されませんでした。これは、Prettier自体には直接的な脆弱性がないことを示しています。

2. 依存関係の問題：
   先ほどの脆弱性の報告は、プロジェクトの他の依存関係（おそらく直接的または間接的に@vercel/funやvercelに関連するもの）に起因していると考えられます。

3. 複雑な依存関係ツリー：
   大規模なプロジェクトや多くの依存関係を持つプロジェクトでは、このような間接的な脆弱性が報告されることがよくあります。これは、あるパッケージが別のパッケージに依存し、そのパッケージがさらに別のパッケージに依存するという具合に、依存関係のツリーが複雑になるためです。

4. 特定のバージョンの問題：
   脆弱性は特定のバージョンのパッケージに存在することが多いです。新しいプロジェクトでは最新バージョンがインストールされるため、既知の脆弱性が含まれていない可能性が高くなります。

5. プロジェクト固有の問題：
   先ほどの脆弱性の報告は、そのプロジェクト特有の依存関係構造に起因している可能性が高いです。

結論として、Prettier自体には問題がなく、プロジェクトの他の依存関係（おそらく@vercel/funやvercel関連）に脆弱性があると考えられます。これらの脆弱性を解決するには、問題のあるパッケージを直接または間接的に使用している部分を特定し、可能であればそれらを更新するか、代替策を検討する必要があります。

このような状況は開発において珍しくありません。重要なのは、報告された脆弱性の重大度を評価し、プロジェクトにとってのリスクを判断することです。そして、適切な対策（更新、パッチの適用、代替パッケージの使用など）を講じることが大切です。

**Reply by siso001** [reply]
自分も前に同じことが起きたんですけど、色々コマンド打っても治らなかったですね...
この問題で開発に困ったことはあまりないので無視していましたけど、こういうプロジェクトだと後々面倒なことが起きそうなのでそうはいかないですよね😓

**Reply by kagomen** [reply]
@siso001 

多分複雑なフレームワークなんかをほとんど使ったことがないからだと思うんですが、自分は今まで`npm audit fix`すればちゃんと直ってたので、今回直らなくてびっくりしましたｗ

一般的には、脆弱性が解消されるのを期待して開発進めたりするんですかね？👀
vercelくらいの規模だったらすぐ修正入りそうな感じもありますよね（てきとう😂）

**Comment by kagomen** [comment]
試しにLGTMでaxiosをnpm installしてみた

```sh
npm i axios

added 9 packages, and audited 632 packages in 3s

151 packages are looking for funding
  run `npm fund` for details

5 vulnerabilities (1 low, 2 moderate, 2 high)

To address all issues, run:
  npm audit fix

Run `npm audit` for details.
```

prettierの時と同じ表示がされた

## わかったこと
脆弱性に関するテキストは、常に表示される！
今インストールしたものに対してではない
なので、何に対して脆弱性が出てるかをちゃんと確認すること！（英語読めない→なんとなくしか読まないのが良くない）

**Comment by kagomen** [comment]
## 結論

主に@vercel/funとその依存関係に問題がある

---

はい、このnpm audit reportを詳しく解説いたします。

1. debug 4.0.0 - 4.3.0の脆弱性:
   - 問題: Regular Expression Denial of Service (ReDoS)
   - 影響: debugパッケージのバージョン4.0.0から4.3.0
   - 場所: node_modules/@vercel/fun/node_modules/debug
   - 修正: npm audit fixで修正可能
   - 依存関係: @vercel/funパッケージが影響を受けるdebugのバージョンに依存している

2. semver 7.0.0 - 7.5.1の脆弱性:
   - 重大度: 高
   - 問題: Regular Expression Denial of Service (ReDoS)
   - 影響: semverパッケージのバージョン7.0.0から7.5.1
   - 場所: node_modules/@vercel/fun/node_modules/semver
   - 修正: npm audit fixで修正可能
   - 依存関係: @vercel/funパッケージが影響を受けるsemverのバージョンに依存している

3. tar <6.2.1の脆弱性:
   - 重大度: 中
   - 問題: tarファイル解析時のフォルダ数の検証不足によるDoS（サービス拒否攻撃）
   - 影響: tarパッケージのバージョン6.2.1未満
   - 場所: node_modules/tar
   - 修正: npm audit fixで修正可能

4. 依存関係の連鎖:
   - vercel >=32.3.1 が @vercel/fun の脆弱性のあるバージョンに依存している
   - @vercel/fun が debug, semver, tar の脆弱性のあるバージョンに依存している

5. 総合的な脆弱性の数:
   - 合計5つの脆弱性（低1、中2、高2）

6. 推奨される対応:
   - npm audit fix を実行して問題の解決を試みる

この報告から、主に@vercel/funとその依存関係に問題があることがわかります。これらの脆弱性は主にRegular Expression Denial of Service（ReDoS）とDoS攻撃に関連しており、攻撃者が特別に細工された入力を使用してシステムリソースを枯渇させる可能性があります。

npm audit fixを実行しても問題が解決しない場合、これらのパッケージの更新やプロジェクトの依存関係の見直しが必要になる可能性があります。特に、vercelと@vercel/funパッケージの使用方法と必要性を再評価することが重要かもしれません。

**Reply by kagomen** [reply]
Regular Expression Denial of Service (ReDoS) とDoS攻撃について、より詳しく説明いたします。

1. Regular Expression Denial of Service (ReDoS):

ReDoSは、悪意のある入力を使って正規表現の処理を異常に長時間実行させ、システムリソースを枯渇させる攻撃です。

具体的な例:
- 攻撃者が、正規表現エンジンの弱点を突く特別な文字列を入力します。
- この文字列は、正規表現のバックトラッキングを大量に発生させます。
- 結果として、単純な処理が指数関数的に時間がかかるようになります。

例えば、以下のような正規表現があるとします：
`^(a+)+$`

この正規表現に対して、攻撃者が "aaaaaaaaaaaaaaaaX" のような文字列を入力すると、バックトラッキングが大量に発生し、処理時間が異常に長くなる可能性があります。

2. DoS (Denial of Service) 攻撃:

DoS攻撃は、システムやネットワークリソースを枯渇させ、正常なサービス提供を妨害する攻撃です。

tarパッケージの脆弱性の場合:
- 攻撃者が特別に細工したtarファイルを作成します。
- このファイルには、大量のフォルダが含まれています。
- tarパッケージがこのファイルを解析しようとすると、フォルダの数の検証が不十分なため、大量のメモリを消費したり、処理時間が異常に長くなったりします。
- 結果として、システムのリソースが枯渇し、他の正常な処理が妨げられます。

これらの攻撃の影響:
1. サーバーの反応が遅くなる、または完全に応答しなくなる。
2. CPUやメモリの使用率が異常に高くなる。
3. 他のユーザーへのサービス提供が妨げられる。
4. 最悪の場合、システムがクラッシュする。

これらの攻撃は、直接的なデータ漏洩やシステム侵害につながるものではありませんが、サービスの可用性を著しく損なう可能性があります。そのため、これらの脆弱性を修正することは、安定したサービス運用のために重要です。

**Reply by kagomen** [reply]
一般的に開発者は、DoS攻撃くらうほどのサービスじゃないから今回の脆弱性に関しては気にしなくて良い、みたいな判断をしたりしなかったりするのかな？

**Reply by kagomen** [reply]
実際に、多くの開発者がそのような判断をすることがあります。しかし、これは必ずしも適切な判断とは言えません。以下に理由を説明します：

1. セキュリティの重要性:
   - 小規模なプロジェクトや個人的なサービスであっても、セキュリティは常に重要です。
   - 攻撃者は時として小さな目標を選ぶことがあり、「目立たない」からといって安全とは限りません。

2. 将来の拡張性:
   - 現在は小規模でも、将来的にプロジェクトが成長する可能性があります。
   - 初期段階からセキュリティを考慮することで、後々の問題を防ぐことができます。

3. 信頼性と評判:
   - セキュリティ問題は、たとえ小規模なプロジェクトでも、開発者の信頼性や評判に影響を与える可能性があります。

4. 間接的な影響:
   - 脆弱性のあるコードが他のプロジェクトに再利用された場合、より大きな問題につながる可能性があります。

5. 攻撃の自動化:
   - 多くの攻撃は自動化されており、規模に関係なくスキャンされる可能性があります。

6. 法的責任:
   - データ保護法や規制によっては、セキュリティ対策の実施が義務付けられている場合があります。

7. 良好な習慣の形成:
   - 常にセキュリティを意識することは、開発者として重要なスキルです。

適切なアプローチ:

1. リスク評価:
   - 脆弱性の深刻度とプロジェクトへの影響を評価する。

2. コストベネフィット分析:
   - 修正にかかる労力と潜在的なリスクを比較する。

3. 段階的な対応:
   - 重要度の高い脆弱性から順に対処する。

4. 定期的な監査:
   - npm auditなどのツールを定期的に実行し、新しい脆弱性をチェックする。

5. ドキュメント化:
   - 既知の脆弱性と対応策（または非対応の理由）を記録する。

6. 継続的な学習:
   - セキュリティベストプラクティスに関する知識を常にアップデートする。

結論として、脆弱性を完全に無視するのではなく、プロジェクトの状況に応じて適切に評価し、対応を決定することが重要です。小規模なプロジェクトであっても、基本的なセキュリティ対策を講じることは良い習慣であり、長期的には有益です。

**Comment by kazzyfrog** [comment]
みた感じ、Cloudflare Pages にデプロイする際に、
インストールしたパッケージの依存関係、かと思います！

Cloudflare Pagesで、Next.jsのSSR（サーバーサイドレンダリング）を動かすのは、割と最近できるようになったので、、、
様子見ながら、まだ実用的じゃなかったら、リリース時にvercel？に移してもいいかもですね！👍

**Reply by kazzyfrog** [reply]
ただ、GitHubのセキュリティアラートは鳴っていないのと、
snykでセキュリティチェックしても、そこまで問題報告されていない感じ、、ですかね！

- [@cloudflare/next-on-pages](https://snyk.io/advisor/npm-package/@cloudflare/next-on-pages)
    - [@vercel/fun](https://snyk.io/advisor/npm-package/@vercel/fun)
- [eslint-plugin-next-on-pages](https://snyk.io/advisor/npm-package/eslint-plugin-next-on-pages)

**Reply by kagomen** [reply]
@kazzyfrog 

調査ありがとうございます🙇🙇

> 様子見ながら、まだ実用的じゃなかったら、リリース時にvercel？に移してもいいかもですね！👍

そうですね！
「実用で使うにはまだ不安が多い」みたいな言葉って、こういうことなんですね！👀

そして、snykっていうのがあるんですね
とても勉強になります……！

**Comment by kagomen** [comment]
## メモ

snyk：脆弱性発見ツール？
https://qiita.com/kannkyo/items/913055157c9105d02926

時間あるとき調べたい！
